dd-trace-core/src/main/java/datadog/trace/core/scopemanager/ContinuableScopeManager.java

```diff
@@ -205,21 +256,131 @@ public String toString() {
     }
   }
 
+  static final class ScopeStack {
+    ContinuableScope[] stack = new ContinuableScope[16];
+    int topPos = -1;
+
+    /**
+     * top - accesses the top of the ScopeStack making sure the Scope on-top is still active If the
+     * top scope isn't active, then the stack is popped back to the top-most active Scope
+     */
+    final ContinuableScope top() {
+      int priorTopPos = this.topPos;
+      if (priorTopPos == -1) return null;
+
+      // localizing & clamping stackPos to enable ArrayBoundsCheck elimination
+      // only bothering to do this here because of the loop below
+      ContinuableScope[] stack = this.stack;
+      priorTopPos = Math.min(priorTopPos, stack.length);
+
+      // Peel first iteration
+      ContinuableScope topScope = stack[priorTopPos];
+      if (topScope.alive()) return topScope;
+
+      // null out top position, it is no longer alive
+      stack[topPos] = null;
+
+      for (int curPos = topPos - 1; curPos >= 0; --curPos) {
+        ContinuableScope curScope = stack[curPos];
+        if (curScope.alive()) {
+          // save the position for next time
+          topPos = curPos;
+
+          if (topPos < stack.length >> 2) {
+            stack = Arrays.copyOf(stack, stack.length >> 1);
```

Comment:
this is an assignment to a local and is never reflected onto the instance
```suggestion
            this.stack = Arrays.copyOf(stack, stack.length >> 1);
```